src/rescharstring.rs

```diff
@@ -0,0 +1,151 @@
+//! BER-encoded Restricted Character String types.
+//!
+//! This is an internal module. It’s public types are re-exported by the
+//! parent.
+
+use std::marker::PhantomData;
+use super::{decode, encode};
+use super::OctetString;
+use super::tag::Tag;
+
+
+//------------ LimitedCharSet ------------------------------------------------
+
+/// Trait for the various Restricted Character String types defined in X.680
+pub trait LimitedCharSet {
+
+    /// Implementations must specify their own natural tag
+    const TAG: Tag;
+
+    /// Implementations must verify the contents
+    fn is_allowed<I: Iterator<Item=u8>>(i: &mut I) -> bool;
```

Comment:
Since we have to have `CharsetError` anyway, this method could be named `check` and return `Result<(), CharsetError>`, allowing to simple question mark out later.

Reply:
Then we need to do additional boilerplate in every implementation of LimitedCharSet, so leaving this as discussed.